src/Workspaces/Core/Portable/VirtualChars/AbstractVirtualCharService.cs

```diff
@@ -0,0 +1,100 @@
+using System.Linq;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.VirtualChars
+{
+    internal abstract class AbstractVirtualCharService : IVirtualCharService
+    {
+        protected abstract ImmutableArray<VirtualChar> TryConvertToVirtualCharsWorker(SyntaxToken token);
+
+        public ImmutableArray<VirtualChar> TryConvertToVirtualChars(SyntaxToken token)
+        {
+            // We don't process any strings that contain diagnostics in it.  That means that we can 
+            // trust that all the string's contents (most importantly, the escape sequences) are well
+            // formed.
+            if (token.ContainsDiagnostics)
+            {
+                return default;
+            }
+
+            var result = TryConvertToVirtualCharsWorker(token);
+
+#if DEBUG
+            // Do some invariant checking to make sure we processed the string token the same
+            // way the C# and VB compilers did.
+            if (!result.IsDefault)
+            {
+                // Ensure that we properly broke up the token into a sequence of characters that
+                // matches what the compiler did.
+                var expectedValueText = token.ValueText;
+                var actualValueText = result.CreateString();
+                Debug.Assert(expectedValueText == actualValueText);
+
+                if (result.Length > 0)
+                {
+                    var currentVC = result[0];
+                    Debug.Assert(currentVC.Span.Start > token.SpanStart, "First span has to start after the start of the string token (including its delimeter)");
+                    Debug.Assert(currentVC.Span.Start == token.SpanStart + 1 || currentVC.Span.Start == token.SpanStart + 2, "First span should start on the second or third char of the string.");
+                    for (var i = 1; i < result.Length; i++)
+                    {
+                        var nextVC = result[i];
+                        Debug.Assert(currentVC.Span.End == nextVC.Span.Start, "Virtual character spans have to be touching.");
+                        currentVC = nextVC;
+                    }
+
+                    var lastVC = result.Last();
+                    Debug.Assert(lastVC.Span.End == token.Span.End - 1, "Last span has to end right before the end of hte string token (including its trailing delimeter).");
+                }
+            }
+#endif
+
+            return result;
+        }
+
+        /// <summary>
+        /// Helper to convert simple string literals that escape quotes by doubling them.  This is 
+        /// how normal VB literals and c# verbatim string literals work.
+        /// </summary>
+        protected ImmutableArray<VirtualChar> TryConvertSimpleDoubleQuoteString(
+            SyntaxToken token, string startDelimeter, string endDelimeter)
+        {
+            Debug.Assert(!token.ContainsDiagnostics);
+
+            var tokenText = token.Text;
```

Comment:
Why `.Text` here, but `.ValueText` above?

Reply:
Very intentional.  We use .ValueText because we want to know that the *interpretted* VirtualChar **chars** sum up to the .ValueText. However, we want to use .Text here becaue we want to ensure that the uninterpretted VirtualChar **spans** completely span it and hold for certain invariants.

In a very real sense VirtualChar is the bridge API between .Text and .ValueText.  The Spans of the VirtualChars come from us breaking up .Text into pieces.  And each VirtualChar has a char that is then each char of .ValueText in order.

Does that make sense?

Reply:
Makes sense. Thanks for the explanation

Reply:
@CyrusNajmabadi Could you document this in the code, so that future you can remember,

Reply:
@AdamSpeight2008 I will add this to the documentation for VirtualChar

Reply:
Added to docs on hte interface.